From e35435b0a11e6c61c8c43b0cf8dc65a563b4a670 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Thu, 10 Apr 2025 13:51:21 +0200
Subject: [PATCH] test-sd-device: limit the number of iterations when testing
 device parent/child functions

The test "hangs" and times out on some arm64 machines. It actually works as
expected, but the machine has 2016 children under /sys/devices/system/memory/,
and the tests do a double loop over this, which is slow enough to hit the 120 s
limit. Add a limit on the number of iterations.

Another option would be to exclude "memory" subsystem. But we may have other
subsystems which have the same problem in the future, so I think it'll be more
robust to not try to limit the fix to a specific subsystem.

(cherry picked from commit 74cb65e45fbf3468cf6b522e4b4fa568d95f12c6)
---
 src/libsystemd/sd-device/test-sd-device.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/libsystemd/sd-device/test-sd-device.c b/src/libsystemd/sd-device/test-sd-device.c
index 620615b6bb..aa235cf8d0 100644
--- a/src/libsystemd/sd-device/test-sd-device.c
+++ b/src/libsystemd/sd-device/test-sd-device.c
@@ -456,6 +456,8 @@ static void check_parent_match(sd_device_enumerator *e, sd_device *dev) {
 
 TEST(sd_device_enumerator_add_match_parent) {
         _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        /* Some devices have thousands of children. Avoid spending too much time in the double loop below. */
+        unsigned iterations = 200;
         int r;
 
         assert_se(sd_device_enumerator_new(&e) >= 0);
@@ -473,6 +475,9 @@ TEST(sd_device_enumerator_add_match_parent) {
                 const char *syspath;
                 sd_device *parent;
 
+                if (iterations-- == 0)
+                        break;
+
                 assert_se(sd_device_get_syspath(dev, &syspath) >= 0);
 
                 r = sd_device_get_parent(dev, &parent);
@@ -501,6 +506,8 @@ TEST(sd_device_enumerator_add_match_parent) {
 
 TEST(sd_device_get_child) {
         _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        /* Some devices have thousands of children. Avoid spending too much time in the double loop below. */
+        unsigned iterations = 3000;
         int r;
 
         assert_se(sd_device_enumerator_new(&e) >= 0);
@@ -534,6 +541,9 @@ TEST(sd_device_get_child) {
                 FOREACH_DEVICE_CHILD_WITH_SUFFIX(parent, child, suffix) {
                         const char *s;
 
+                        if (iterations-- == 0)
+                                return;
+
                         assert_se(child);
                         assert_se(suffix);
 
